main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
+    ///
+    pub def findRight(f: a -> Bool, a: Array[a]): Option[a] =
+        match findIndexOfRight(x -> f(x), a) {
+            case None => None
+            case Some(i) => Some(a[i] as & Pure)
+        }
+
+    ///
+    /// Returns an array of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns `[]` if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Array[Int32] & Impure =
+        rangeHelper(b,e)
+
+    /// 
+    /// Helper for `range`.
+    ///
+    def rangeHelper(b: Int32, e: Int32): Array[Int32] & Impure =
+        if (b >= e)
+            []
+        else {
+            let f = x -> x + b;
+            init(f, e-b)
+        }
+
+    ///
+    /// Returns an array with the element `x` repeated `n` times.
+    ///
+    /// Returns `[]` if `n <= 0`.
+    ///
+    pub def repeat(x: a, n: Int32): Array[a] & Impure =
+        if (n > 0)
+            new(x, n)
+        else
+            []
+
+    ///
+    /// Alias for `scanLeft`.
+    ///
+    pub def scan(f: (b, a) ->b & e, s: b, xs: Array[a]): Array[b] & Impure = scanLeft(f, s, xs)
+
+    ///
+    /// Accumulates the result of applying `f` to `a` going left to right.
+    ///
+    /// That is, the result is of the form: `[s , f(s, x1), f(f(s, x1), x2),  ...]`.
+    ///
+    pub def scanLeft(f: (b, a) -> b & e, s: b, a: Array[a]): Array[b] & Impure =
+        let len = length(a) + 1;
+        let b = new(s, len);
+        scanLeftHelper(f, s, a, b, len, 1);
+        b
+
+    ///
+    /// Helper for `scanLeft`.
+    ///
+    def scanLeftHelper(f: (b, a) -> b & e, s: b, a: Array[a], b: Array[b], len: Int32, i: Int32) : Unit & Impure =
+        if (i >= len)
+            ()
+        else {
+            let s1 = f(s, a[i-1]);
+            b[i] = s1;
+            scanLeftHelper(f, s1, a, b, len, i+1)
+        }
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going right to left.
+    ///
+    /// That is, the result is of the form: `[..., f(xn-1, f(xn, s)), f(xn, s), s]`.
+    ///
+    pub def scanRight(f: (a, b) -> b & e, s: b, a: Array[a]): Array[b] & Impure =
+        let len = length(a);
+        let b = new(s, len+1);
+        scanRightHelper(f, s, a, b, len-1);
+        b
+
+    ///
+    /// Helper for `scanRight`.
+    ///
+    def scanRightHelper(f: (a, b) -> b & e, s: b, a: Array[a], b: Array[b], i: Int32) : Unit & Impure =
+        if (i < 0)
+            ()
+        else {
+            let s1 = f(a[i], s);
+            b[i] = s1;
+            scanRightHelper(f, s1, a, b, i-1)
+        }
+
+    ///
+    /// Returns the result of applying `f` to every element in `a`.
+    ///
+    /// The result is a new array.
+    ///
+    pub def map(f: a -> b & e, a: Array[a]): Array[b] & Impure =
+        let len = length(a);
+        init(i -> f(a[i]), len)
+
+    ///
+    /// Apply `f` to every element in array `a`. Array `a` is mutated.
+    ///
+    pub def mapInPlace(f: a -> a, a: Array[a]): Unit & Impure =
```

Comment:
I have been toying with the idea that purely functional and mutating functions should have different names.

For example, for `Set` we have `map` but for `MutSet` we have `transform`:
https://github.com/flix/flix/blob/master/main/src/library/MutMap.flix#L232

This way, there is no confusion between `map` and `transform`.

On the other hand, a naming scheme using a suffix like `InPlace` might be more scalable and intuitive. 

So I am not sure. I don't particularly like "InPlace" but I don't have any better suggestion at the moment.

I would like for us to avoid the mistake that Scala made where functions that *share the same name may behave differently*. But I think we already agree on that.